axum/src/middleware/map_request.rs

```diff
@@ -0,0 +1,462 @@
+}
+
+impl<F, S, I, T> Clone for MapRequest<F, S, I, T>
+where
+    F: Clone,
+    I: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            f: self.f.clone(),
+            inner: self.inner.clone(),
+            state: Arc::clone(&self.state),
+            _extractor: self._extractor,
+        }
+    }
+}
+
+macro_rules! impl_service {
+    (
+        [$($ty:ident),*], $last:ident
+    ) => {
+        #[allow(non_snake_case, unused_mut)]
+        impl<F, Fut, S, I, B, $($ty,)* $last> Service<Request<B>> for MapRequest<F, S, I, ($($ty,)* $last,)>
+        where
+            F: FnMut($($ty,)* $last) -> Fut + Clone + Send + 'static,
+            $( $ty: FromRequestParts<S> + Send, )*
+            $last: FromRequest<S, B> + Send,
+            Fut: Future + Send + 'static,
+            Fut::Output: IntoMapRequestResult<B> + Send + 'static,
+            I: Service<Request<B>, Error = Infallible>
+                + Clone
+                + Send
+                + 'static,
+            I::Response: IntoResponse,
+            I::Future: Send + 'static,
+            B: Send + 'static,
+            S: Send + Sync + 'static,
+        {
+            type Response = Response;
+            type Error = Infallible;
+            type Future = ResponseFuture;
+
+            fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+                self.inner.poll_ready(cx)
+            }
+
+            fn call(&mut self, req: Request<B>) -> Self::Future {
+                let not_ready_inner = self.inner.clone();
+                let mut ready_inner = std::mem::replace(&mut self.inner, not_ready_inner);
+
+                let mut f = self.f.clone();
+                let state = Arc::clone(&self.state);
+
+                let future = Box::pin(async move {
+                    let (mut parts, body) = req.into_parts();
+
+                    $(
+                        let $ty = match $ty::from_request_parts(&mut parts, &state).await {
+                            Ok(value) => value,
+                            Err(rejection) => return rejection.into_response(),
+                        };
+                    )*
+
+                    let req = Request::from_parts(parts, body);
+
+                    let $last = match $last::from_request(req, &state).await {
+                        Ok(value) => value,
+                        Err(rejection) => return rejection.into_response(),
+                    };
+
+                    match f($($ty,)* $last).await.into_map_request_result() {
+                        Ok(req) => {
+                            ready_inner.call(req).await.into_response()
+                        }
+                        Err(res) => {
+                            res
+                        }
+                    }
+                });
+
+                ResponseFuture {
+                    inner: future
+                }
+            }
+        }
+    };
+}
+
+impl_service!([], T1);
```

Comment:
At this point, should `all_the_tuples!` just be updated to this pattern instead of the array-free one? Idk whether I've said this elsewhere but my idea to get rid of the array syntax in these macros turned out to not really work.

Reply:
Yes I think it should. Its only `HandleError` that uses the old version https://github.com/tokio-rs/axum/blob/main/axum/src/error_handling/mod.rs#L207. I'll do that separately.